lnwallet/chancloser/rbf_coop_test.go

```diff
@@ -0,0 +1,1647 @@
+
+	return state
+}
+
+type shutdownExpect struct {
+	isInitiator   bool
+	allowSend     bool
+	recvShutdown  bool
+	finalBalances fn.Option[ShutdownBalances]
+}
+
+func (r *rbfCloserTestHarness) expectShutdownEvents(expect shutdownExpect) {
+	r.T.Helper()
+
+	r.chanObserver.On("FinalBalances").Return(expect.finalBalances)
+
+	if expect.isInitiator {
+		r.chanObserver.On("NoDanglingUpdates").Return(expect.allowSend)
+	} else {
+		r.chanObserver.On("NoDanglingUpdates").Return(
+			expect.allowSend,
+		).Maybe()
+	}
+
+	// When we're receiving a shutdown, we should also disable incoming
+	// adds.
+	if expect.recvShutdown {
+		r.chanObserver.On("DisableIncomingAdds").Return(nil)
+	}
+
+	// If a close is in progress, this is an RBF iteration, the link has
+	// already been cleaned up, so we don't expect any assertions, other
+	// than to check the closing state.
+	if expect.finalBalances.IsSome() {
+		return
+	}
+
+	r.chanObserver.On("DisableOutgoingAdds").Return(nil)
+
+	r.chanObserver.On(
+		"MarkShutdownSent", mock.Anything, expect.isInitiator,
+	).Return(nil)
+
+	chanPoint := r.env.ChanPoint
+
+	r.daemonAdapters.On("DisableChannel", chanPoint).Return(nil)
+}
+
+func (r *rbfCloserTestHarness) expectFinalBalances(b fn.Option[ShutdownBalances]) {
+	r.chanObserver.On("FinalBalances").Return(b)
+}
+
+func (r *rbfCloserTestHarness) expectIncomingAddsDisabled() {
+	r.T.Helper()
+
+	r.chanObserver.On("DisableIncomingAdds").Return(nil)
+}
+
+type msgMatcher func([]lnwire.Message) bool
+
+func singleMsgMatcher[M lnwire.Message](f func(M) bool) msgMatcher {
+	return func(msgs []lnwire.Message) bool {
+		if len(msgs) != 1 {
+			return false
+		}
+
+		wireMsg := msgs[0]
+
+		msg, ok := wireMsg.(M)
+		if !ok {
+			return false
+		}
+
+		if f == nil {
+			return true
+		}
+
+		return f(msg)
+	}
+}
+
+func (r *rbfCloserTestHarness) expectMsgSent(matcher msgMatcher) {
+	r.T.Helper()
+
+	if matcher == nil {
+		r.daemonAdapters.On(
+			"SendMessages", r.peerPub, mock.Anything,
+		).Return(nil)
+	} else {
+		r.daemonAdapters.On(
+			"SendMessages", r.peerPub, mock.MatchedBy(matcher),
+		).Return(nil)
+	}
+}
+
+func (r *rbfCloserTestHarness) expectFeeEstimate(absoluteFee btcutil.Amount,
+	numTimes int) {
+
+	r.T.Helper()
+
+	// TODO(roasbeef): mo assertions for dust case
+
+	r.feeEstimator.On(
+		"EstimateFee", mock.Anything, mock.Anything, mock.Anything,
+		mock.Anything,
+	).Return(absoluteFee, nil).Times(numTimes)
+}
+
+func (r *rbfCloserTestHarness) expectFaliure(err error) {
+	r.T.Helper()
+
+	errorMatcher := func(e error) bool {
+		return errors.Is(e, err)
+	}
+
+	r.errReporter.On("ReportError", mock.MatchedBy(errorMatcher)).Return()
+}
+
+func (r *rbfCloserTestHarness) expectNewCloseSig(
+	localScript, remoteScript []byte, fee btcutil.Amount,
+	closeBalance btcutil.Amount, sequence uint32) {
+
+	r.T.Helper()
+
+	// TODO(roasbeef): not working? need custom checker?
+	/*functionalOpt := mock.FunctionalOptions(
+		lnwallet.WithCustomSequence(sequence),
+	)*/
```

Comment:
🧐

Reply:
Heh this was that bug fix for taproot chans. With the latest version though, we'll use lock time instead.